mempool/clist_mempool.go

```diff
@@ -537,14 +607,17 @@ func (mem *CListMempool) ReapMaxBytesMaxGas(maxBytes, maxGas int64) types.Txs {
 
 	// TODO: we will get a performance boost if we have a good estimate of avg
 	// size per tx, and set the initial capacity based off of that.
-	// txs := make([]types.Tx, 0, cmtmath.MinInt(mem.txs.Len(), max/mem.avgTxSize))
-	txs := make([]types.Tx, 0, mem.txs.Len())
-	for e := mem.txs.Front(); e != nil; e = e.Next() {
-		memTx := e.Value.(*mempoolTx)
-
-		txs = append(txs, memTx.tx)
+	// txs := make([]types.Tx, 0, cmtmath.MinInt(mem.Size(), max/mem.avgTxSize))
+	txs := make([]types.Tx, 0, mem.Size())
+	mem.reapIter.Reset(mem.lanes)
```

Comment:
There's a risk of race conditions here, as shown in the [failed test](https://github.com/cometbft/cometbft/actions/runs/10686803248/job/29622932068?pr=3638) `TestBroadcastTxSync`. We assumed that Reap methods are only called by BlockExecutor, but this is not true. In particular, `ReapMaxTxs` is called by the RPC endpoint `UnconfirmedTxs`. A solution is to just create a new iterator on every Reap call, instead of reusing only one.

Reply:
Fixed by #3996